File: src/main/java/net/reflact/client/map/MapTextureManager.java
package net.reflact.client.map;

import net.minecraft.client.MinecraftClient;
import net.minecraft.client.texture.NativeImage;
import net.minecraft.client.texture.NativeImageBackedTexture;
import net.minecraft.util.Identifier;

public class MapTextureManager {

    private static final int MAP_SIZE = 128; // Standard map size
    private final NativeImage nativeImage;
    private NativeImageBackedTexture dynamicTexture; // Remove final
    private final Identifier textureId;
    private boolean dirty = false;
    
    // Singleton instance
    public static final MapTextureManager INSTANCE = new MapTextureManager();

    // Cache last player position to avoid pointless updates
    private int lastPlayerX = Integer.MAX_VALUE;
    private int lastPlayerZ = Integer.MAX_VALUE;
    
    private boolean initialized = false;

    private MapTextureManager() {
        this.nativeImage = new NativeImage(NativeImage.Format.RGBA, MAP_SIZE, MAP_SIZE, false);
        // Fill with transparent or background color initially
        for (int x = 0; x < MAP_SIZE; x++) {
            for (int y = 0; y < MAP_SIZE; y++) {
                this.nativeImage.setColor(x, y, 0xFF000000); // Black opaque
            }
        }
        
        // NativeImageBackedTexture(NativeImage) creation moved to init() to avoid early RenderSystem access
        this.textureId = Identifier.of("reflact", "world_map_dynamic");
    }
    
    public void init() {
        if (initialized) return;
        this.dynamicTexture = new NativeImageBackedTexture(() -> "Reflact Map", this.nativeImage);
        MinecraftClient.getInstance().getTextureManager().registerTexture(this.textureId, this.dynamicTexture);
        initialized = true;
    }

    public void update(MinecraftClient client) {
        if (!initialized) init();

        if (client.player == null || client.world == null) return;
        
        int px = client.player.getBlockPos().getX();
        int pz = client.player.getBlockPos().getZ();
        
        int radius = MAP_SIZE / 2;
        
        for (int x = 0; x < MAP_SIZE; x++) {
            for (int z = 0; z < MAP_SIZE; z++) {
                int worldX = px - radius + x;
                int worldZ = pz - radius + z;
                
                int y = client.world.getTopY(net.minecraft.world.Heightmap.Type.WORLD_SURFACE, worldX, worldZ);
                net.minecraft.util.math.BlockPos pos = new net.minecraft.util.math.BlockPos(worldX, y - 1, worldZ);
                net.minecraft.block.BlockState state = client.world.getBlockState(pos);
                
                int color = state.getMapColor(client.world, pos).color;
                
                if (color == 0) {
                    color = 0xFF000000;
                } else {
                    // Convert MapColor (ARGB) to ABGR for NativeImage
                    int r = (color >> 16) & 0xFF;
                    int g = (color >> 8) & 0xFF;
                    int b = color & 0xFF;
                    int a = 0xFF;
                    color = (a << 24) | (b << 16) | (g << 8) | r;
                }
                
                nativeImage.setColor(x, z, color);
            }
        }

        this.dynamicTexture.upload();
    }
    
    public void setMapData(byte[] colors, int width, int height) {
        if (!initialized) init();
        // Assuming colors is RGBA
        // NativeImage expects RGBA? Or ABGR?
        // NativeImage methods usually take int color.
        // We can loop.
        
        if (width != MAP_SIZE || height != MAP_SIZE) {
             // Handle resize or mismatch? 
             // For now assume 128x128 match
        }
        
        for (int i = 0; i < colors.length / 4; i++) {
            int index = i * 4;
            int x = i % width;
            int y = i / width;
            
            // Reconstruct int color
            // If server sent B, G, R, A
            int b = colors[index] & 0xFF;
            int g = colors[index+1] & 0xFF;
            int r = colors[index+2] & 0xFF;
            int a = colors[index+3] & 0xFF;
            
            // NativeImage setPixelColor usually expects ABGR packed int on little endian?
            // Or RGBA? Minecraft uses ABGR internally usually.
            // Let's try 0xAABBGGRR
            int color = (a << 24) | (b << 16) | (g << 8) | r;
            
            if (x < MAP_SIZE && y < MAP_SIZE) {
                this.nativeImage.setColor(x, y, color);
            }
        }
        dirty = true;
    }

    public Identifier getTextureId() {
        return textureId;
    }
    
    public void close() {
        // Do nothing for singleton
    }
}

File: src/main/kotlin/net/reflact/client/map/MapTextureManager.kt
package net.reflact.client.map

import net.minecraft.client.MinecraftClient
import net.minecraft.client.texture.NativeImage
import net.minecraft.client.texture.NativeImageBackedTexture
import net.minecraft.util.Identifier
import net.minecraft.world.Heightmap

class MapTextureManager private constructor() {

    private val nativeImage: NativeImage = NativeImage(NativeImage.Format.RGBA, MAP_SIZE, MAP_SIZE, false)
    private lateinit var dynamicTexture: NativeImageBackedTexture
    val textureId: Identifier = Identifier.of("reflact", "world_map_dynamic")
    private var dirty = false
    private var initialized = false

    private var lastServerUpdate: Long = 0

    init {
        // Fill with transparent or background color initially
        for (x in 0 until MAP_SIZE) {
            for (y in 0 until MAP_SIZE) {
                this.nativeImage.setColor(x, y, 0xFF000000.toInt()) // Black opaque
            }
        }
    }

    fun init() {
        if (initialized) return
        this.dynamicTexture = NativeImageBackedTexture({ "Reflact Map" }, this.nativeImage)
        MinecraftClient.getInstance().textureManager.registerTexture(this.textureId, this.dynamicTexture)
        initialized = true
    }

    fun update(client: MinecraftClient) {
        if (!initialized) init()

        // If we received server map data recently (within 6 seconds), don't overwrite with client scan
        if (System.currentTimeMillis() - lastServerUpdate < 6000) {
            this.dynamicTexture.upload()
            return
        }

        if (client.player == null || client.world == null) return

        val px = client.player!!.blockPos.x
        val pz = client.player!!.blockPos.z

        val radius = MAP_SIZE / 2

        for (x in 0 until MAP_SIZE) {
            for (z in 0 until MAP_SIZE) {
                val worldX = px - radius + x
                val worldZ = pz - radius + z

                val y = client.world!!.getTopY(Heightmap.Type.WORLD_SURFACE, worldX, worldZ)
                val pos = net.minecraft.util.math.BlockPos(worldX, y - 1, worldZ)
                val state = client.world!!.getBlockState(pos)

                val color = state.getMapColor(client.world, pos).color

                if (color == 0) {
                    // Void / Black - Use transparent or dark gray
                    nativeImage.setColor(x, z, 0xFF000000.toInt())
                } else {
                    // MapColor.color is usually integer ARGB.
                    // NativeImage expects ABGR (Little Endian).
                    // We need to swap Red and Blue.

                    val alpha = 255 // Force opaque
                    val red = (color shr 16) and 0xFF
                    val green = (color shr 8) and 0xFF
                    val blue = color and 0xFF

                    // ABGR
                    val abgr = (alpha shl 24) or (blue shl 16) or (green shl 8) or red
                    nativeImage.setColor(x, z, abgr)
                }
            }
        }

        this.dynamicTexture.upload()
    }

    fun setMapData(colors: ByteArray, width: Int, height: Int) {
        if (!initialized) init()
        this.lastServerUpdate = System.currentTimeMillis()

        if (width != MAP_SIZE || height != MAP_SIZE) {
            // Handle resize or mismatch?
            // For now assume 128x128 match
        }

        for (i in 0 until colors.size / 4) {
            val index = i * 4
            val x = i % width
            val y = i / width

            // Reconstruct int color
            // If server sent B, G, R, A
            val b = colors[index].toInt() and 0xFF
            val g = colors[index + 1].toInt() and 0xFF
            val r = colors[index + 2].toInt() and 0xFF
            val a = colors[index + 3].toInt() and 0xFF

            // NativeImage setPixelColor usually expects ABGR packed int on little endian?
            // Let's try 0xAABBGGRR
            val color = (a shl 24) or (b shl 16) or (g shl 8) or r

            if (x < MAP_SIZE && y < MAP_SIZE) {
                this.nativeImage.setColor(x, y, color)
            }
        }
        dirty = true
    }

    fun close() {
        // Do nothing for singleton
    }

    companion object {
        const val MAP_SIZE = 128 // Standard map size
        val INSTANCE = MapTextureManager()
    }
}
